src/main/scala/fi/oph/koski/raportit/SuoritustietojenTarkistus.scala

```diff
@@ -0,0 +1,274 @@
+      valmiitYhteistenTutkinnonOsatLkm = yhteisetTutkinnonOsat.size,
+      pakollisetYhteistenTutkinnonOsienOsaalueidenLkm = pakolliset(yhteistenTutkinnonOsienOsaAlueet).size,
+      valinnaistenYhteistenTutkinnonOsienOsaalueidenLKm = valinnaiset(yhteistenTutkinnonOsienOsaAlueet).size,
+      tunnustettujaTukinnonOsanOsaalueitaValmiissaTutkinnonOsanOsalueissaLkm = tunnustetut(yhteistenTutkinnonOsienOsaAlueet).size,
+      rahoituksenPiirissäTutkinnonOsanOsaalueitaValmiissaTutkinnonOsanOsaalueissaLkm = rahoituksenPiirissä(yhteistenTutkinnonOsienOsaAlueet).size,
+      tunnustettujaYhteistenTutkinnonOsienValmiistaOsistaLkm = tunnustetut(yhteisetTutkinnonOsat).size,
+      rahoituksenPiirissäTunnustetuistaYhteisenTutkinnonOsistaLkm = (tunnustetut andThen rahoituksenPiirissä) (yhteisetTutkinnonOsat).size,
+      suoritettujenYhteistenTutkinnonOsienYhteislaajuus = yhteislaajuus(yhteisetTutkinnonOsat),
+      suoritettujenYhteistenTutkinnonOsienOsaalueidenYhteislaajuus = yhteislaajuus(yhteistenTutkinnonOsienOsaAlueet),
+      pakollistenYhteistenTutkinnonOsienOsaalueidenYhteislaajuus = (pakolliset andThen yhteislaajuus) (yhteistenTutkinnonOsienOsaAlueet),
+      valinnaistenYhteistenTutkinnonOsienOsaalueidenYhteisLaajuus = (valinnaiset andThen yhteislaajuus) (yhteistenTutkinnonOsienOsaAlueet)
+    )
+  }
+
+  private def päätasonSuoritustenTilat(päätasonsuoritukset: Seq[RPäätasonSuoritusRow]) = {
+    päätasonsuoritukset.map(ps => vahvistusPäivätToTila(ps.vahvistusPäivä)).mkString(",")
+  }
+
+  private[raportit] def vahvistusPäivätToTila(vahvistusPäivä: Option[Date]) = vahvistusPäivä match {
+      case Some(päivä) if (isTulevaisuudessa(päivä)) =>  s"Kesken, Valmistuu ${päivä.toLocalDate}"
+      case Some(_) =>  "Valmis"
+      case _ => "Kesken"
+    }
+
+  private def ammatillisetTutkinnonOsatJaOsasuorituksetFrom(osasuoritukset: Seq[ROsasuoritusRow]) = {
+    osasuoritukset.filter(os =>
+      isAnyOf(os,
+        isAmmatillisenLukioOpintoja,
+        isAmmatillisenKorkeakouluOpintoja,
+        isAmmatillinenMuitaOpintoValmiuksiaTukeviaOpintoja,
+        isAmmatillinenTutkinnonOsaaPienempiKokonaisuus,
+        isAmmatillisenTutkinnonOsa,
+        isAmmatillisenYhteisenTutkinnonOsienOsaalue(_, osasuoritukset)
+      ))
+  }
+
+  private def isAnyOf(osasuoritus: ROsasuoritusRow, fs: (ROsasuoritusRow => Boolean)*) = fs.exists(f => f(osasuoritus))
+
+  private val tutkinnonosatvalinnanmahdollisuusKoodiarvot = Seq("1,", "2")
+
+  private val isAmmatillisenTutkinnonOsa: ROsasuoritusRow => Boolean = osasuoritus => {
+    !(tutkinnonosatvalinnanmahdollisuusKoodiarvot.contains(osasuoritus.koulutusmoduuliKoodiarvo)) &
+      (osasuoritus.suorituksenTyyppi == "ammatillisentutkinnonosa") &
+      (tutkinnonOsanRyhmä(osasuoritus, "1"))
+  }
+
+  private val isArvioinniton: ROsasuoritusRow => Boolean = osasuoritus => isAnyOf(osasuoritus,
+    isAmmatillisenLukioOpintoja,
```

Comment:
Idea varoittelee "suspicicous forward reference":istä. Jos käyttää val:eja jotka kutsuvat toisiaan objectissa parempi laittaa ne järjestykseen siten, että kutsujat tulevat viimeiseksi. Näin vältytään oudoilta bugeilta jotka johtuu objectin initialisaatiojärjestyksestä.